### PR TITLE
NCGenerics: synthesize Copyable/Escapable decls

### PR DESCRIPTION
**Pull request by kavon**

### CODE REVIEW

File: include/swift/SIL/SILType.h [hunk 1]
```diff
@@ -763,9 +763,13 @@ class SILType {
...
-  /// Returns true if this type is a first class move only type or a move only
-  /// wrapped type.
-  bool isMoveOnly() const;
+  /// Returns true if this type is a noncopyable type. Otherwise, if the type
+  /// satisfies \c isMoveOnlyWrapped(), then it returns true iff \c orWrapped
+  /// is true. That is,
+  ///
+  /// orWrapped == false -->  isNoncopyable
+  /// orWrapped == true  -->  isNoncopyable || isMoveOnlyWrapped
+  bool isMoveOnly(bool orWrapped=true) const;
```
Comment (eeckstein) [inline include/swift/SIL/SILType.h:772]: At this occasion, can you rename this API to `isCopyable` (with the invert meaning)?

**Review by eeckstein — COMMENTED** [review]
Does this actually change something in the mangling scheme which requires changes in the demangler?

File: include/swift/SIL/SILType.h (hunk 1 above)
Comment (kavon) [reply include/swift/SIL/SILType.h]: Something that's wrapped in a SILMoveOnlyWrapper isn't necessarily noncopyable. It's just not implicitly copyable, so perhaps it should really imply that instead.

**Comment by kavon** [comment]
Mangler changes are needed to mangle the uses of `~Copyable` into generic signatures, but that's now moving to another PR for now (Sorry for changing the focus of this one).

**Comment by kavon** [comment]
@swift-ci please test

**Comment by kavon** [comment]
@swift-ci please test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Review by ahoppen — COMMENTED** [review]
Parser changes look good to me. It probably needs some adjustment for code completion though. It would also be good to change the `complete_wihout_constraint_type.swift` test to check that we only suggest `Copyable` after `~` and not all the types by changing it to the following (which currently passes).

```swift
// RUN: %batch-code-completion

struct FileDescriptor: ~#^COMPLETE^# {}
// COMPLETE: Begin completions, 1 item
// COMPLETE-DAG: Decl[Protocol]/OtherModule[Swift]/IsSystem: Copyable[#Copyable#]; name=Copyable
```

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

**Comment by kavon** [comment]
@swift-ci smoke test

File: lib/Sema/TypeCheckInvertible.cpp [hunk 2]
```diff
@@ -468,13 +468,9 @@ ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
...
     //
-    // I'm currently unsure what happens when rebuilding a module from its
-    // interface, so this might not be unreachable code just yet.
-    if (SWIFT_ENABLE_EXPERIMENTAL_NONCOPYABLE_GENERICS &&
-        file->getKind() != FileUnitKind::Synthesized) {
-      llvm_unreachable("when can this actually happen??");
-    }
-
+    // There are various other kinds of SourceFiles, like SIL, which instead
+    // get their conformances here instead.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckInvertible.cpp:472]: Why don't source file nominals also get their conformance from here then?

**Review by slavapestov — APPROVED** [review]

**Pull request merged**